hamilton/function_modifiers/delayed.py

```diff
@@ -0,0 +1,150 @@
+import enum
+import inspect
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from hamilton import settings
+from hamilton.function_modifiers.base import (
+    DynamicResolver,
+    InvalidDecoratorException,
+    NodeTransformLifecycle,
+)
+
+
+class ResolveAt(enum.Enum):
+    STATIC = "static"
+    COMPILE = "config"
+    RUNTIME = "runtime"
+    NEVER = "never"
+
+
+VALID_PARAM_KINDS = [inspect.Parameter.POSITIONAL_OR_KEYWORD, inspect.Parameter.KEYWORD_ONLY]
+
+
+def extract_and_validate_params(fn: Callable) -> Tuple[List[str], List[str]]:
+    """Gets the parameters from a function, while validating that
+    the function has *only* named arguments.
+
+    :param fn: Function to extract parameters from
+    :return: List of parameter names
+    :raises InvalidDecoratorException: If the function has any non kwargs-friendly arguments
+    """
+    invalid_params = []
+    required_params = []
+    optional_params = []
+    sig = inspect.signature(fn)
+    for key, value in inspect.signature(fn).parameters.items():
+        if value.kind not in VALID_PARAM_KINDS:
+            invalid_params.append(key)
+        else:
+            if value.default is not value.empty:
+                optional_params.append(key)
+            else:
+                required_params.append(key)
+    if invalid_params:
+        raise InvalidDecoratorException(
+            f"Configuration-parsing functions can only except keyword-friendly arguments. "
+            f"Instead got signature: {sig}"
+        )
+    return required_params, optional_params
+
+
+class dynamic(DynamicResolver):
```

Comment:
I don't think dynamic is the write word here.

Perhaps having it under `@config.inject` would make the most sense?

Reply:
`@config.resolve` ?

Reply:
Hmm... Some ideas:
```python
config.inject()
config.resolve()
config.when_available()
resolve_at(time=COMPILE)
resolve_at(time=CONFIG_AVAILABLE)
resolve_when(CONFIG_AVAILABLE)
delay_resolution(until=CONFIG_AVAILABLE)
_(when=CONFIG_AVAILABLE)
```

So far I like putting it under `config` -- maybe `resolve` as you said? Not a fan of `inject`, especially since we're already using it for something else.

Reply:
In my head the behavior would be something like:
```python
# signature
@config.resolve(func)
def my_func(...)
```
If flag is off:
 - resolve() -> {}

if flag is on and parameters not in config:
 - resolve() -> {}
 
if flag is on and parameters are in config:
 - resolve() -> {my_func}

Reply:
> *

I think it should hard break if it's not going to work. If you need specifics in the settings to get it to work, then I'd rather fail fast...

Reply:
yes let's keep it pretty constrained now -- and then relax if needed.